File: GPS/GPS.m
classdef GPS < matlab.System & coder.ExternalDependency
        % GPS- this block sends its inputs to active X-Plane simulation using shared memory. Note: in order for this block to work, special X-Plane plugin needs to be installed
    
        properties(Access = public)
        
        end
            
        properties(DiscreteState)
    
        end
    
        properties(Access = private)
    
        end
    
        methods(Access = protected)
            function setupImpl(obj)
                    coder.cinclude('GPS.h');
                    coder.ceval('setup_GPS');
            end
    
            function [lat,lon,elevation] = stepImpl(obj)
                lat = double(0);
                lon = double(0);
                elevation = double(0);
                coder.cinclude('GPS.h');
                coder.ceval('update_GPS');
                lat = coder.ceval('get_lat');
                lon = coder.ceval('get_lon');
                elevation = coder.ceval('get_elevation');
            end
    
            function resetImpl(obj)
                
            end
            
            function releaseImpl(obj)
                coder.cinclude('GPS.h');
                coder.ceval('close_GPS');
            end
        end
            
        methods (Static)
            function name = getDescriptiveName()
                name = 'GPS';
            end
            function b = isSupportedContext(context)
                b = context.isCodeGenTarget('rtw');
            end
            function updateBuildInfo(buildInfo, context)
            % Update the build-time buildInfo
                blockRoot = fileparts(mfilename('fullpath'));
                buildInfo.addIncludePaths({blockRoot});
                buildInfo.addIncludeFiles({'GPS.h'});
                buildInfo.addSourcePaths({blockRoot});
                buildInfo.addSourceFiles({'GPS.c'});
            end
        end
    end

File: GPS/GPS.c
#include "GPS.h"


void setup_GPS()
{
    GPS_Mapping = OpenFileMapping(
        FILE_MAP_ALL_ACCESS,   // read/write access
        FALSE,                 // do not inherit the name
        GPS_Name);               // name of mapping object
    
    if (GPS_Mapping == NULL) {
        GPS_isMapped = false;
        return;
    }

    GPS_Buf = (LPTSTR)MapViewOfFile(GPS_Mapping, // handle to map object
        FILE_MAP_ALL_ACCESS,  // read/write permission
        0,
        0,
        BUF_SIZE);
    
    if (GPS_Buf == NULL) {
        GPS_isMapped = false;
        return;
    }
    GPS_isMapped = true;
}


double get_lat()
{
    if (GPS_isMapped) {
        return GPS_struct.lat;
    } else {
        return 0;
    }
}

double get_lon()
{
    if (GPS_isMapped) {
        return GPS_struct.lon;
    } else {
        return 0;
    }
}

double get_elevation()
{
    if (GPS_isMapped) {
        return GPS_struct.elevation;
    } else {
        return 0;
    }
}

void update_GPS() {
    if (GPS_isMapped) {
        memcpy(&GPS_struct, GPS_Buf, sizeof(GPS_struct));
    }
}

void close_GPS()
{
    if (GPS_isMapped) {
        
        UnmapViewOfFile(GPS_Buf);
        CloseHandle(GPS_Mapping);
    }
}